plugins/out_azure_blob/azure_blob_store.c

```diff
@@ -0,0 +1,454 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "azure_blob_store.h"
+#include <fluent-bit/flb_output_plugin.h>
+#include <fluent-bit/flb_fstore.h>
+#include <fluent-bit/flb_time.h>
+#include <fcntl.h>
+
+/*
+ * Simple and fast hashing algorithm to create keys in the local buffer
+ */
+static flb_sds_t gen_store_filename(const char *tag)
+{
+    int c;
+    unsigned long hash = 5381;
+    unsigned long hash2 = 5381;
+    flb_sds_t hash_str;
+    flb_sds_t tmp;
+    struct flb_time tm;
+
+    /* get current time */
+    flb_time_get(&tm);
+
+    /* compose hash */
+    while ((c = *tag++)) {
+        hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+    }
+    hash2 = (unsigned long) hash2 * tm.tm.tv_sec * tm.tm.tv_nsec;
+
+    /* flb_sds_printf allocs if the incoming sds is not at least 64 bytes */
+    hash_str = flb_sds_create_size(64);
+    if (!hash_str) {
+        flb_errno();
+        return NULL;
+    }
+    tmp = flb_sds_printf(&hash_str, "%lu-%lu", hash, hash2);
+    if (!tmp) {
+        flb_errno();
+        flb_sds_destroy(hash_str);
+        return NULL;
+    }
+    hash_str = tmp;
+
+    return hash_str;
+}
+
+
+/* Retrieve a candidate buffer file using the tag */
+struct azure_blob_file *azure_blob_store_file_get(struct flb_azure_blob *ctx, const char *tag,
+                                                    int tag_len)
+{
+    struct mk_list *head;
+    struct mk_list *tmp;
+    struct flb_fstore_file *fsf = NULL;
+    struct azure_blob_file *azure_blob_file;
+    int found = 0;
+
+    /*
+     * Based in the current ctx->stream_name, locate a candidate file to
+     * store the incoming data using as a lookup pattern the content Tag.
+     */
+    mk_list_foreach_safe(head, tmp, &ctx->stream_active->files) {
+        fsf = mk_list_entry(head, struct flb_fstore_file, _head);
+
+        /* skip and warn on partially initialized chunks */
+        if (fsf->data == NULL) {
+            flb_plg_warn(ctx->ins, "BAD: found flb_fstore_file with NULL data reference, tag=%s, file=%s, will try to delete", tag, fsf->name);
+            flb_fstore_file_delete(ctx->fs, fsf);
+        }
+
+        if (fsf->meta_size != tag_len) {
+            fsf = NULL;
+            continue;
+        }
+
+        /* skip locked chunks */
+        azure_blob_file = fsf->data;
+        if (azure_blob_file->locked == FLB_TRUE) {
+            flb_plg_debug(ctx->ins, "File '%s' is locked, skipping", fsf->name);
```

Comment:
what this message means for the end-user ? I would suggest something that cannot be interpreted as an error